api/src/main/java/io/grpc/GlobalInterceptors.java

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright 2022 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc;
+
+import com.google.common.collect.ImmutableList;
```

Comment:
Since `java.util.Collections.unmodifiableList` creates a wrapper around the same existing List such that the wrapper cannot be used to modify it. The returned list is only truly immutable if nobody holds a reference to the original list, if not original list can still be modified which makes it unsafe.

Is there any reason why we prefer `Collections.unmodifiableList` over `ImmutableList`?

Reply:
> Is there any reason why we prefer Collections.unmodifiableList over ImmutableList?

Eric mentioned "We aren't currently using the immutable collections in code used on Android." Since the guava dependency is already included I don't know what it means but for Android it might be good to avoid these classes.

Regarding absolute immutability - do you know what the actual requirement from our point of view is? Most of the times what is required is the collection we return from our getters should not be used by our callers to modify the list. But what you are saying is that without absolute immutability (the way you described) the collection(s) we return can have different contents at different points in time. But I think our setters are setting these collections as unmodifiable/immutable in which case it is a moot point?